modules/src/ics04_channel/handler/send_packet.rs

```diff
@@ -0,0 +1,267 @@
+use crate::ics04_channel::packet::{PacketResult, Sequence};
+use crate::ics04_channel::{context::ChannelReader, error::Error, error::Kind, packet::Packet};
+use crate::ics24_host::identifier::{ChannelId, PortId};
+use crate::Height;
+
+#[derive(Clone, Debug)]
+pub struct SendPacketResult {
+    pub port_id: PortId,
+    pub channel_id: ChannelId,
+    pub seq: Sequence,
+    pub seq_number: Sequence,
+    pub timeout_height: Height,
+    pub timeout_timestamp: u64,
+    pub data: Vec<u8>,
+}
+
+pub fn send_packet(ctx: &dyn ChannelReader, packet: Packet) -> HandlerResult<PacketResult, Error> {
+    let mut output = HandlerOutput::builder();
+
+    if packet.sequence.is_zero() {
+        return Err(Kind::ZeroPacketSequence.into());
+    }
+    if packet.timeout_height.is_zero() && packet.timeout_timestamp == 0 {
+        return Err(Kind::ZeroPacketTimeout.into());
+    }
+    if packet.data.is_empty() {
+        return Err(Kind::ZeroPacketData.into());
+    }
+
+    let source_channel_end = ctx
+        .channel_end(&(packet.source_port.clone(), packet.source_channel.clone()))
+        .ok_or_else(|| {
+            Kind::ChannelNotFound(packet.source_port.clone(), packet.source_channel.clone())
+                .context(packet.source_channel.clone().to_string())
+        })?;
+
+    if source_channel_end.state_matches(&State::Closed) {
+        return Err(Kind::ChannelClosed(packet.source_channel).into());
+    }
+
+    // TODO: Capability Checked in Ics20 as well. Check Difs.
+    let _channel_cap = ctx.authenticated_capability(&packet.source_port)?;
+
+    let counterparty = Counterparty::new(
+        packet.destination_port.clone(),
+        Some(packet.destination_channel.clone()),
+    );
+
+    if !source_channel_end.counterparty_matches(&counterparty) {
+        return Err(Kind::InvalidPacketCounterparty(
+            packet.destination_port.clone(),
+            packet.destination_channel,
+        )
+        .into());
+    }
+
+    let connection_end = ctx
+        .connection_end(&source_channel_end.connection_hops()[0])
+        .ok_or_else(|| Kind::MissingConnection(source_channel_end.connection_hops()[0].clone()))?;
+
+    let client_id = connection_end.client_id().clone();
+
+    let client_state = ctx
+        .client_state(&client_id)
+        .ok_or_else(|| Kind::MissingClientState(client_id.clone()))?;
+
+    // prevent accidental sends with clients that cannot be updated
+    if client_state.is_frozen() {
+        return Err(Kind::FrozenClient(connection_end.client_id().clone()).into());
+    }
+
+    // check if packet height timeouted on the receiving chain
+
+    let latest_height = client_state.latest_height();
+    let packet_height = packet.timeout_height;
+
+    if !packet.timeout_height.is_zero() && packet_height.cmp(&latest_height).eq(&Ordering::Less) {
+        return Err(Kind::LowPacketHeight(latest_height, packet.timeout_height).into());
+    }
+
+    //check if packet timestamp timeouted on the receiving chain
```

Comment:
We don't really check the receiving chain. This (like the timeout height above) is a weak check, but if the latest client consensus state we have for the receiving chain is newer (higher timestamp) then for sure we have a timeout. 
```suggestion
    //check if the packet timeout timestamp is newer than the timestamp of the latest consensus of the receiving chain
```